### PR TITLE
Add Validate() methods to each message

### PR DESCRIPTION
**Pull request by koden-km**
#### What change does this introduce?

All messages will have `Validate()` methods added to implement the `ValidatableMessage` interface.

#### Why make this change?

Some bugs slipped through tests recently because messages were not constructed correctly. Dogma has a message validation feature that the engine will use if messages implement the `ValidatableMessage` interface.

#### Is there anything you are unsure about?

* Should the `messages.IsValidBusinessDate()` be un-exported and copied into both `commands` and `events` packages?
* If `messages.IsValidBusinessDate()` is staying as exported, should `domain.mustValidateDate()` use the exported `messages.IsValidBusinessDate()` instead of doing the parse logic again?

#### What issues does this relate to?

#97

### CODE REVIEW

**Comment by jmalloc** [comment]
> * Should the `messages.IsValidBusinessDate()` be un-exported and copied into both `commands` and `events` packages?

> * If `messages.IsValidBusinessDate()` is staying as exported, should `domain.mustValidateDate()` use the exported `messages.IsValidBusinessDate()` instead of doing the parse logic again?

Perhaps the validation functions should be in their own package `validation` or something like that. I don't think it makes sense to keep it exported in `messages` if its used for validating things that aren't messages.

One question though, is `domain.mustValidateDate()` still needed when the messages are guaranteed to be validated?

**Comment by koden-km** [comment]
Yeah ok, a `validation` package could make sense for that.

I guess that `mustValidateDate()` is no longer needed. It currently doesn't add anything extra. Do you think there is any need to validate the date relative to a specific time? I'm guessing no because you can schedule things for the future and also for "today" which is basically instantly the past.

**Comment by jmalloc** [comment]
> I guess that `mustValidateDate()` is no longer needed. It currently doesn't add anything extra. Do you think there is any need to validate the date relative to a specific time? I'm guessing no because you can schedule things for the future and also for "today" which is basically instantly the past.

Nah I don't think so. That sounds more like some kind of logic, not just a "well-formed check" which is the kind of validation we are doing here.

If `mustValidateDate()` is no longer needed I would probably suggest making an internal message-only validation package at `messages/internal/validation`.

**Comment by koden-km** [comment]
I have moved the `IsValidateBusinessDate()` to an internal package and removed the older `mustValidateDate()` function.

**Comment by koden-km** [comment]
I have added the marshal functions as discussed on Slack.

**Comment by koden-km** [comment]
This is now ready for review. This branch has been updated with the latest changes about removing business date.

**Review by danilvpetrov — APPROVED** [review]

File: messages/commands/account.go [hunk 1]
```diff
@@ -83,3 +84,72 @@ func (m DebitAccount) MessageDescription() string {
 		m.AccountID,
 	)
 }
+
+// Validate returns a non-nil error if the message is invalid.
+func (m OpenAccountForNewCustomer) Validate() error {
+	if m.CustomerID == "" {
+		return errors.New("OpenAccountForNewCustomer needs a valid customer ID")
```
Comment (jmalloc) [inline messages/commands/account.go:91]:
```suggestion
		return errors.New("OpenAccountForNewCustomer must not have an empty customer ID")
```

File: messages/commands/account.go [hunk 2]
```diff
@@ -83,3 +84,72 @@ func (m DebitAccount) MessageDescription() string {
...
+		return errors.New("CreditAccount needs a valid transaction ID")
+	}
+	if m.AccountID == "" {
+		return errors.New("CreditAccount needs a valid account ID")
+	}
+	if m.TransactionType == "" {
+		return errors.New("CreditAccount needs a valid transaction type")
+	}
+	if m.Amount < 1 {
+		return errors.New("CreditAccount needs a valid amount")
```
Comment (jmalloc) [inline messages/commands/account.go:133]:
```suggestion
		return errors.New("CreditAccount must have a positive amount")
```

File: messages/events/dailydebitlimit.go [hunk 3]
```diff
@@ -53,3 +55,57 @@ func (m DailyDebitLimitExceeded) MessageDescription() string {
...
+		return errors.New("DailyDebitLimitConsumed needs a valid amount")
+	}
+	if !validation.IsValidBusinessDate(m.Date) {
+		return errors.New("DailyDebitLimitConsumed needs a valid date")
+	}
+	if m.TotalDebitsForDay < 1 {
+		return errors.New("DailyDebitLimitConsumed needs a valid total debits for day")
+	}
+	if m.DailyLimit < 0 {
+		return errors.New("DailyDebitLimitConsumed needs a valid daily limit")
```
Comment (jmalloc) [inline messages/events/dailydebitlimit.go:80]:
```suggestion
		return errors.New("DailyDebitLimitConsumed must have a non-negative daily limit")
```

File: messages/marshal.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package messages
+
+import "time"
+
+// businessDateFormat is the format used for business dates in messages.
+const businessDateFormat = "2006-01-02"
+
+// MarshalDate marshals a time value to a date string.
```
Comment (jmalloc) [inline messages/marshal.go:8]: Has this not been rebased off `master`. I thought these functions where gone?

**Review by jmalloc — CHANGES_REQUESTED** [review]

File: messages/marshal.go (hunk 4 above)
Comment (koden-km) [reply messages/marshal.go:8]: It had been. Maybe they got re-added during the merge.

Comment (koden-km) [reply messages/marshal.go:8]: Actually these seem like they are still needed for validation. I could move them into the `internal/validation/validation.go` file though?

Comment (jmalloc) [reply messages/marshal.go:8]: Are they really both needed?

I think you could just make `IsValidBusinessDate()` do what it needs to do.

File: messages/commands/account.go (hunk 1 above)
Comment (koden-km) [reply messages/commands/account.go]: Suggested applied.

File: messages/commands/account.go (hunk 2 above)
Comment (koden-km) [reply messages/commands/account.go]: Suggested applied.

File: messages/marshal.go (hunk 4 above)
Comment (koden-km) [reply messages/marshal.go]: The `marshal.go` file has been removed and `validation.go` updated to do all it needs to validate a date.

File: messages/events/dailydebitlimit.go (hunk 3 above)
Comment (koden-km) [reply messages/events/dailydebitlimit.go]: I accidentally worded this one different. Let me know if you prefer your suggestion:
I ended up with this: `DailyDebitLimitConsumed must not have a negative daily limit`

Comment (jmalloc) [reply messages/events/dailydebitlimit.go]: Your wording is better.

File: messages/commands/dailydebitlimit.go [hunk 5]
```diff
@@ -27,3 +28,24 @@ func (m ConsumeDailyDebitLimit) MessageDescription() string {
...
+	if m.AccountID == "" {
+		return errors.New("ConsumeDailyDebitLimit must not have an empty account ID")
+	}
+	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
+		return errors.New("ConsumeDailyDebitLimit must have a valid debit type")
+	}
+	if m.Amount < 1 {
+		return errors.New("ConsumeDailyDebitLimit must have a positive amount")
+	}
+	if m.Date == "" {
```
Comment (jmalloc) [inline messages/commands/dailydebitlimit.go:46]:
```suggestion
	if !validation.IsValidDate(m.Date) {
```

File: messages/commands/dailydebitlimit.go [hunk 6]
```diff
@@ -27,3 +28,24 @@ func (m ConsumeDailyDebitLimit) MessageDescription() string {
...
+// Validate returns a non-nil error if the message is invalid.
+func (m ConsumeDailyDebitLimit) Validate() error {
+	if m.TransactionID == "" {
+		return errors.New("ConsumeDailyDebitLimit must not have an empty transaction ID")
+	}
+	if m.AccountID == "" {
+		return errors.New("ConsumeDailyDebitLimit must not have an empty account ID")
+	}
+	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
+		return errors.New("ConsumeDailyDebitLimit must have a valid debit type")
```
Comment (jmalloc) [inline messages/commands/dailydebitlimit.go:41]: This one throws away the original error altogether, probably better split into separate checks.

File: messages/events/dailydebitlimit.go [hunk 7]
```diff
@@ -53,3 +55,57 @@ func (m DailyDebitLimitExceeded) MessageDescription() string {
...
+
+// Validate returns a non-nil error if the message is invalid.
+func (m DailyDebitLimitConsumed) Validate() error {
+	if m.TransactionID == "" {
+		return errors.New("DailyDebitLimitConsumed must not have an empty transaction ID")
+	}
+	if m.AccountID == "" {
+		return errors.New("DailyDebitLimitConsumed must not have an empty account ID")
+	}
+	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
```
Comment (jmalloc) [inline messages/events/dailydebitlimit.go:67]: Split.

File: messages/events/dailydebitlimit.go [hunk 8]
```diff
@@ -53,3 +55,57 @@ func (m DailyDebitLimitExceeded) MessageDescription() string {
...
+
+// Validate returns a non-nil error if the message is invalid.
+func (m DailyDebitLimitExceeded) Validate() error {
+	if m.TransactionID == "" {
+		return errors.New("DailyDebitLimitExceeded must not have an empty transaction ID")
+	}
+	if m.AccountID == "" {
+		return errors.New("DailyDebitLimitExceeded must not have an empty account ID")
+	}
+	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
```
Comment (jmalloc) [inline messages/events/dailydebitlimit.go:94]: Split.

**Review by jmalloc — CHANGES_REQUESTED** [review]

File: messages/commands/dailydebitlimit.go (hunk 5 above)
Comment (koden-km) [reply messages/commands/dailydebitlimit.go]: Fixed missing date validate as suggested.

File: messages/commands/dailydebitlimit.go (hunk 6 above)
Comment (koden-km) [reply messages/commands/dailydebitlimit.go]: Split debit type validation for better error reporting.

File: messages/commands/dailydebitlimit.go [hunk 9]
```diff
@@ -37,13 +38,16 @@ func (m ConsumeDailyDebitLimit) Validate() error {
 	if m.AccountID == "" {
 		return errors.New("ConsumeDailyDebitLimit must not have an empty account ID")
 	}
-	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
-		return errors.New("ConsumeDailyDebitLimit must have a valid debit type")
+	if err := m.DebitType.Validate(); err != nil {
+		return fmt.Errorf("ConsumeDailyDebitLimit must have a valid debit transaction type: %w", err)
+	}
+	if !m.DebitType.IsDebit() {
+		return errors.New("ConsumeDailyDebitLimit must have a valid debit transaction type")
```
Comment (jmalloc) [inline messages/commands/dailydebitlimit.go:45]:
```suggestion
		return fmt.Errorf("ConsumeDailyDebitLimit must have a valid transaction type: %w", err)
	}
	if !m.DebitType.IsDebit() {
		return errors.New("ConsumeDailyDebitLimit must have a debit transaction type")
```

File: messages/events/dailydebitlimit.go [hunk 10]
```diff
@@ -64,8 +64,11 @@ func (m DailyDebitLimitConsumed) Validate() error {
 	if m.AccountID == "" {
 		return errors.New("DailyDebitLimitConsumed must not have an empty account ID")
 	}
-	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
-		return errors.New("DailyDebitLimitConsumed must have a valid debit type")
+	if err := m.DebitType.Validate(); err != nil {
+		return fmt.Errorf("DailyDebitLimitConsumed must have a valid debit transaction type: %w", err)
+	}
+	if !m.DebitType.IsDebit() {
+		return errors.New("DailyDebitLimitConsumed must have a valid debit transaction type")
```
Comment (jmalloc) [inline messages/events/dailydebitlimit.go:71]:
```suggestion
		return fmt.Errorf("DailyDebitLimitConsumed must have a valid transaction type: %w", err)
	}
	if !m.DebitType.IsDebit() {
		return errors.New("DailyDebitLimitConsumed must have a debit transaction type")
```

File: messages/events/dailydebitlimit.go [hunk 11]
```diff
@@ -91,8 +94,11 @@ func (m DailyDebitLimitExceeded) Validate() error {
 	if m.AccountID == "" {
 		return errors.New("DailyDebitLimitExceeded must not have an empty account ID")
 	}
-	if err := m.DebitType.Validate(); err != nil && !m.DebitType.IsDebit() {
-		return errors.New("DailyDebitLimitExceeded must have a valid debit type")
+	if err := m.DebitType.Validate(); err != nil {
+		return fmt.Errorf("DailyDebitLimitExceeded must have a valid debit transaction type: %w", err)
+	}
+	if !m.DebitType.IsDebit() {
+		return errors.New("DailyDebitLimitExceeded must have a valid debit transaction type")
```
Comment (jmalloc) [inline messages/events/dailydebitlimit.go:101]:
```suggestion
		return fmt.Errorf("DailyDebitLimitExceeded must have a valid transaction type: %w", err)
	}
	if !m.DebitType.IsDebit() {
		return errors.New("DailyDebitLimitExceeded must have a debit transaction type")
```

**Review by jmalloc — CHANGES_REQUESTED** [review]
Ok, these are my last suggestions, what do you think?

**Review by jmalloc — APPROVED** [review]

**Pull request merged**